extensions/ql-vscode/src/databases/github-database-module.ts

```diff
@@ -121,15 +139,29 @@ export class GithubDatabaseModule extends DisposableObject {
     }
 
     if (databases.length === 0) {
+      // If the user didn't have an access token, they have already been prompted,
+      // so we should give feedback.
+      if (!hasAccessToken) {
```

Comment:
I think the behavior is correct, but I found the logic a bit hard to follow. It also seems a bit odd that there is a database prompt (the one to connect) in here instead of in `github-database-prompt.ts`.

I also wonder, what if the repo is public - do we then ask for a non-needed token?

Reply:
The prompt for connecting is defined in this file because it is common to both downloading and updating databases. This should make sense in #3079.

If the repo is public, we do ask for an unnecessary token. I can change this to always make a request first, and then only prompt if we get a 404. However, that will make this logic even harder to follow.

Reply:
> However, that will make this logic even harder to follow.

That is probably right, although I am not super sure. I think what makes it complicated
is that we dont want to ask the user to connect and then ask them again to download the DB. If we asked
both questions then it could probably be written something like this:
```
databases = getDatabases(current_credentials)

if (databases.length === 0 && no_credentials) {
  // Ask the user for credentials and try again
  databases = getDatabases(ask_for_credentials)
}

downloadDatabaseFromGitHub(databases)
```

What do you think? How much do we want to prevent the second question?

Reply:
I think I've found a somewhat simpler solution which will make this easier to follow and also skip the prompt for credentials when the repo is public. I'll create a new function `listDatabases` which will:
1. Try retrieving the databases, either with or without credentials depending on what is available
2. If that fails with a 404 and they don't have an access token, ask the user whether they want to connect to GitHub
3. If they do, prompt for credentials and retry the request

This function can then return a result like `{ databases: CodeqlDatabase[], promptedForCredentials: boolean }`. The logic in the module will then be a little bit simpler since the name is `promptedForCredentials` instead of `hasAccessToken`, more closely matching what the intention is.

Reply:
That sounds good, thanks!

Reply:
I've just pushed this change, the logic for that is mostly inside `github-database-api.ts` and the `listDatabases` function.

Reply:
That looks better and also covers the public repo case. Thanks!